Nonrecursive dfs

rrwick opened the pull request:
The previous implementation of depthFirstSearch was recursive and could crash if it got too many levels deep (10s of thousands). This new implementation is iterative and can handle larger graphs.

This PR replaces https://github.com/seqan/seqan/pull/1823, which was made against master, instead of develop.

h-2 commented on include/seqan/graph_algorithms/depth_first_search.h:
```diff
@@ -29,13 +29,15 @@
 //
 // ==========================================================================
-// Author: Tobias Raussch <rausch@embl.de>
+// Authors: Tobias Raussch <rausch@embl.de>, Ryan Wick <rrwick@gmail.com>
```
We usually just have two author lines, could you change this to
```c++
// Author: Tobias Raussch <rausch@embl.de>
// Author: Ryan Wick <rrwick@gmail.com>
```
?

h-2 commented on include/seqan/graph_algorithms/depth_first_search.h:
```diff
@@ -121,28 +95,80 @@ void depthFirstSearch(TPredecessorMap & predecessor,
     typedef typename Value<TPredecessorMap>::Type TPredVal;
+    typedef typename Iterator<Graph<TSpec>, AdjacencyIterator>::Type TAdjacencyIterator;
+
+    enum class Task : uint8_t
```
Task is a little too general as a name. Since it resides in the global seqan namespace, could you give a little more specific name? If this is never used in a publicly documented interface, please also append an undercore to its name to mark it as "internal/private".

h-2 commented on include/seqan/graph_algorithms/depth_first_search.h:
```diff
@@ -29,13 +29,15 @@
 
 #ifndef INCLUDE_SEQAN_GRAPH_ALGORITHMS_DEPTH_FIRST_SEARCH_H_
 #define INCLUDE_SEQAN_GRAPH_ALGORITHMS_DEPTH_FIRST_SEARCH_H_
+#include <vector>
```
please insert a blank linke before this for readability!

h-2 commented on include/seqan/graph_algorithms/strongly_connected_compnents.h:
```diff
@@ -59,6 +59,35 @@ namespace seqan {
 // ----------------------------------------------------------------------------
 
+template <typename TSpec, typename TVertexDescriptor, typename TTokenMap, typename TPredecessorMap, typename TDiscoveryTimeMap, typename TFinishingTimeMap, typename TVal>
+void
```
please insert `inline`

h-2 commented on include/seqan/graph_algorithms/strongly_connected_compnents.h:
```diff
@@ -59,6 +59,35 @@ namespace seqan {
 // Function stronglyConnectedComponents()
 // ----------------------------------------------------------------------------
 
+template <typename TSpec, typename TVertexDescriptor, typename TTokenMap, typename TPredecessorMap, typename TDiscoveryTimeMap, typename TFinishingTimeMap, typename TVal>
```
why was this function moved to a different file?

h-2 commented:
Thank you for you new PR! Could you please look at the minor issues? You also now need to sign the CLAHUB-thing... thanks!

rrwick commented:
Okay, I've made the changes you've suggested. Thanks!

I moved the `_dfsVisit` function to `strongly_connected_compnents.h` because that is now the only file that uses it. It was previously used in the recursive DFS in `depth_first_search.h`, but that file doesn't use it anymore now that I've rewritten `depthFirstSearch`.

And the CLAHub agreement isn't working for me. I fill out the form, but it just says 'We're sorry, but something went wrong.' I've tried it in a couple of different browsers, with the same result.

h-2 commented:
OK, I think this is good, but I am moving it to @rrahn for final review, because I am on vacation.

rrahn commented on include/seqan/graph_algorithms/depth_first_search.h:
```diff
@@ -30,13 +30,17 @@
 #define INCLUDE_SEQAN_GRAPH_ALGORITHMS_DEPTH_FIRST_SEARCH_H_
 
+#include <vector>
+#include <utility>
```
Can you please move these includes to the umbrella header: in include/seqan/graph_algorithms.h

Here: https://github.com/seqan/seqan/blob/master/include/seqan/graph_algorithms.h#L39

rrahn commented:
Ok, only the minor remark with moving the STL includes to the umbrella header. 
Then we'll merge it.

rrwick commented:
Okay - all done!

rrahn commented:
Great job! Thank you for contributing.
